File: backend/models/order.go
package models

import (
	"time"

	"github.com/jinzhu/gorm"
)

type Order struct {
	Id                uint `gorm:"primary_key"`
	UserId            uint `sql:"not null;index:UserId"`
	CreatedAt         time.Time
	UpdatedAt         time.Time
	BrokerAccountId   uint   `sql:"not null;index:BrokerAccountId"`
	BrokerRef         string `sql:"not null;index:BrokerRef"`
	BrokerAccountRef  string `sql:"not null;index:BrokerAccountRef"`
	Type              string
	SymbolId          uint
	OptionSymbolId    uint
	Side              string
	Qty               int
	Status            string
	Duration          string
	Price             float64 `sql:"type:DECIMAL(12,2)"`
	AvgFillPrice      float64 `sql:"type:DECIMAL(12,2)"`
	ExecQuantity      float64 `sql:"type:DECIMAL(12,2)"`
	LastFillPrice     float64 `sql:"type:DECIMAL(12,2)"`
	LastFillQuantity  float64 `sql:"type:DECIMAL(12,2)"`
	RemainingQuantity float64 `sql:"type:DECIMAL(12,2)"`
	CreateDate        time.Time
	TransactionDate   time.Time
	Class             string
	NumLegs           int
	PositionReviewed  string `sql:"not null;type:ENUM('No', 'Yes', 'Error');default:'No'"`
	Symbol            Symbol `json:"symbol"`
	OptionSymbol      Symbol `gorm:"foreignkey:OptionSymbolId" json:"option_symbol"`
	Legs              []OrderLeg
}

type OrderLeg struct {
	Id                uint `gorm:"primary_key"`
	UserId            uint `sql:"not null;index:UserId"`
	CreatedAt         time.Time
	UpdatedAt         time.Time
	OrderId           uint `sql:"not null;index:OrderId"`
	Type              string
	SymbolId          uint
	Side              string
	Qty               int
	Status            string
	Duration          string
	AvgFillPrice      float64 `sql:"type:DECIMAL(12,2)"`
	ExecQuantity      float64 `sql:"type:DECIMAL(12,2)"`
	LastFillPrice     float64 `sql:"type:DECIMAL(12,2)"`
	LastFillQuantity  float64 `sql:"type:DECIMAL(12,2)"`
	RemainingQuantity float64 `sql:"type:DECIMAL(12,2)"`
	CreateDate        time.Time
	TransactionDate   time.Time
	Symbol            Symbol `json:"symbol"`
}

//
// Store a new order.
//
func (t *DB) CreateOrder(order *Order) error {

	// Create order
	t.Create(order)

	// Return happy
	return nil
}

//
// Store a order.
//
func (t *DB) UpdateOrder(order *Order) error {

	// Update entry.
	t.Save(&order)

	// Return happy
	return nil
}

//
// Get orders by User and class and status and reviewed
//
func (t *DB) GetOrdersByUserClassStatusReviewed(userId uint, class string, status string, reviewed string) ([]Order, error) {

	orders := []Order{}

	// Query and get all orders we have not reviewed before.
	t.Debug().Preload("Legs", func(db *gorm.DB) *gorm.DB {
		return db.Order("id asc")
	}).Where("user_id = ? AND class = ? AND status = ? AND position_reviewed = ?", userId, class, status, reviewed).Order("transaction_date asc").Find(&orders)

	// Return happy
	return orders, nil
}

//
// See if we have an order by user and broker ref.
//
func (t *DB) HasOrderByBrokerRefUserId(brokerId string, userId uint) bool {

	// See if we already have this record in our database
	var count int
	order := &Order{}

	// Run query
	t.Where("broker_ref = ? AND user_id = ?", brokerId, userId).First(order).Count(&count)

	if count > 0 {
		return true
	}

	// Return not found.
	return false
}

/* End File */

File: frontend/src/app/models/order.ts
//
// Date: 9/8/2017
// Author(s): Spicer Matthews (spicer@options.cafe)
// Copyright: 2017 Cloudmanic Labs, LLC. All rights reserved.
//

import { OrderLeg } from './order-leg';

export class Order {
  constructor(
    public Id: number,
    public AccountId: string,
    public AvgFillPrice: number,
    public Class: string,
    public CreateDate: string,
    public Duration: string,
    public ExecQuantity: string,
    public LastFillPrice: number,
    public LastFillQuantity: number,
    public NumLegs: number,
    public Price: number,
    public Quantity: number,
    public RemainingQuantity: number,
    public Side: string,
    public Status: string,
    public Symbol: string,
    public TransactionDate: string,
    public Type: string,
    public OptionSymbol: string,
    public Legs: OrderLeg[]
  ){}

  //
  // Build build the data for emitting to the app. 
  //
  public static buildForEmit(data) : Order[] {
    
    let orders = [];
    
    // Sometimes we do not have data yet.
    if(! data)
    {
      return orders;
    }
    
    for(let i = 0; i < data.length; i++)
    {
      // Add in the legs
      let legs = [];
      
      if(data[i].legs)
      {
        for(let k = 0; k < data[i].legs.length; k++)
        {
          legs.push(new OrderLeg(
            data[i].legs[k].class,
            data[i].legs[k].symbol,
            data[i].legs[k].option_symbol, 
            data[i].legs[k].side, 
            data[i].legs[k].quantity, 
            data[i].legs[k].status, 
            data[i].legs[k].duration, 
            data[i].legs[k].avg_fill_price, 
            data[i].legs[k].exec_quantity, 
            data[i].legs[k].last_fill_price, 
            data[i].legs[k].last_fill_quantity, 
            data[i].legs[k].remaining_quantity, 
            data[i].legs[k].create_date, 
            data[i].legs[k].transaction_date          
          ));
        }
      }
      
      // Push the order on
      orders.push(new Order(
          data[i].id,
          data[i].account_id,
          data[i].avg_fill_price,
          data[i].class,
          data[i].create_date,
          data[i].duration,
          data[i].exec_quantity,
          data[i].last_fill_price,
          data[i].last_fill_quantity,
          data[i].num_legs,
          data[i].price,
          data[i].quantity,
          data[i].remaining_quantity,
          data[i].side,
          data[i].status,
          data[i].symbol,
          data[i].transaction_date,
          data[i].type,
          data[i].option_symbol,
          legs));
               
    }
    
    return orders;
        
  } 

}

/* End File */